xarray/core/concat.py

```diff
@@ -24,14 +25,19 @@
     from .dataarray import DataArray
     from .dataset import Dataset
 
+compat_options = Literal[
+    "identical", "equals", "broadcast_equals", "no_conflicts", "override"
+]
+concat_options = Literal["all", "minimal", "different"]
+
 
 @overload
 def concat(
     objs: Iterable["Dataset"],
-    dim: Union[str, "DataArray", pd.Index],
-    data_vars: Union[str, List[str]] = "all",
-    coords: Union[str, List[str]] = "different",
-    compat: str = "equals",
+    dim: Union[Hashable, "DataArray", pd.Index],
+    data_vars: Union[concat_options, List[Hashable]] = "all",
+    coords: Union[concat_options, List[Hashable]] = "different",
+    compat: compat_options = "equals",
```

Comment:
How about removing the Union and using `|` while we're here?

```suggestion
    dim: Hashable | "DataArray" | pd.Index,
    data_vars: concat_options | List[Hashable] = "all",
    coords: concat_options | List[Hashable] = "different",
    compat: compat_options = "equals",
```

Reply:
Yes, I thought `pyupgrade` did this for us; I can have a look why that isn't working if no one gets there first